### PR TITLE
Make harvest wait for LDD to be ready when it is updated.

### LINKED ISSUE #1: verify artifact is pushed to Maven central and setup the ci

### PR DESCRIPTION
**Pull request by tloubrieu-jpl**
### 🤖 AI Assistance Disclosure
- [ ] No AI assistance used
- [ ] AI used for light assistance (e.g., suggestions, refactoring, documentation help, minor edits)
- [ ] AI used for moderate content generation (AI generated some code or logic, but the developer authored or heavily revised the majority)
- [x] AI generated substantial portions of this code

Estimated % of code influenced by AI: _90_ %

## ⚙️ Test Data and/or Report

Tested with integration tests locally (docker compose) and on AWS dev.

## 🤓 Reviewer Checklist

*Reviewers: Please verify the following before approving this pull request.*

### Documentation and PR Content
- [ ] **Documentation:** README, Wiki, or inline documentation (Sphinx, Javadoc, Docstrings) have been updated to reflect these changes.
- [ ] **Issue Traceability:** The PR is linked to a valid GitHub Issue
- [ ] **PR Title:** The PR title is "user-friendly" clearly identifying what is being fixed or the new feature being added, that if you saw it in the Release Notes for a tool, you would be able to get the gist of what was done.

### Security & Quality
- [ ] **SonarCloud:** Confirmed no new High or Critical security findings.
- [ ] **Secrets Detection:** Verified that the Secrets Detection scan passed and no sensitive information (keys, tokens, PII) is exposed.
- [ ] **Code Quality:** Code follows organization style guidelines and best practices for the specific language (e.g., PEP 8, Google Java Style).

### Testing & Validation
- [ ] **Test Accuracy:** Verified that test data is accurate, representative of real-world PDS4 scenarios, and sufficient for the logic being tested.
- [ ] **Coverage:** Automated tests cover new logic and edge cases.
- [ ] **Local Verification:** (If applicable) Successfully built and ran the changes in a local or staging environment.

### Maintenance
- [ ] **Backward Compatibility:** Confirmed that these changes do not break existing downstream dependencies or API contracts (or that breaking changes are clearly documented).

### CODE REVIEW

**Comment by tloubrieu-jpl** [comment]
@jordanpadams @al-niessner can you review this PR so that I can run the integration test again on the dev account together with Jordan's fix in PR https://github.com/NASA-PDS/registry-common/pull/287 ? Thanks

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 1]
```diff
@@ -120,6 +129,28 @@ public void loadOnly(File lddFile, String lddFileName, String namespace, Instant
...
+      
+ 
+      LddVersions info = dao.getLddInfoNoCache(namespace);
+
+      /* 
+      Collection<String> ids = new ArrayList<String>();
+      ids.add("insight:Observation_Information/insight:stop_local_mean_solar_time"); 
+      List<Tuple> returnedLDDFields = dao.getDataTypes(ids);
+      log.info("Returned LDD field: " + returnedLDDFields.toString());
+      */
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: Please always provide a comment indicating _why_ something is commented-out.

And if you can't come up with a reason why, then just nuke the commented-out code completely, please!

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 2]
```diff
@@ -120,6 +129,28 @@ public void loadOnly(File lddFile, String lddFileName, String namespace, Instant
...
+      
+      
+      while (info.isEmpty()) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: Wouldn't this loop just wait until _some_ LDD info exists for the namespace? It does not wait until the the newly-loaded `lddFileName` appears, contradicting the commenton line 134. Maybe it should be?
```java
while (!info.files.contains(lddFileName)) {
    …
}
```
but with a timeout. Without that, the race may still exist. The code can proceeed as soon as an old LDD is visible.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 3]
```diff
@@ -120,6 +129,28 @@ public void loadOnly(File lddFile, String lddFileName, String namespace, Instant
...
+        Thread.sleep(1000);
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java:136]: Thinking further, if indexing fails, or LDD info never appears, this'll just loop (hang) forever. Can we do some timeout and attempt counting?
```java
int attempts = 0;
int maxAttempts = 60;

while (!info.files.contains(lddFileName) && attempts++ < maxAttempts) {
    Thread.sleep(1000);
    ...
}

if (!info.files.contains(lddFileName)) {
    throw new IOException("Timed out waiting for LDD " + lddFileName + " to be indexed");
}
```

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java [hunk 4]
```diff
@@ -26,6 +28,9 @@
 
 class SearchImpl implements Search {
   final SearchRequest.Builder craftsman = new SearchRequest.Builder();
+
+  private Random rand = new Random();
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: The `rand` is only used in `buildListLddsNoCache` so it doesn't need to be in instances of objects of the class; it can just be local to that method.

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java [hunk 5]
```diff
@@ -88,6 +93,21 @@ public Search buildListLdds(String namespace) {
...
   }
+
+@Override
+public Search buildListLddsNoCache(String namespace) {
+    // add a random must_not clause to prevent Elasticsearch from returning cached results. This is needed to ensure that the new LDD is indexed before any other LDDs are loaded (e.g., from other threads) and potentially overwrite the new LDD with an older one. TODO: remove this workaround after we migrate to managed opensearch
+    BoolQuery.Builder journeyman = new BoolQuery.Builder()
+        .must(this.matchQuery("class_ns", "registry").build(),
+            this.matchQuery("class_name", "LDD_Info").build(),
+            this.matchQuery("attr_ns", namespace).build())
+        .mustNot(this.matchQuery("xxxxxxxxxnotexisting", String.valueOf(rand.nextInt())).build());
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: A few concerns here:

- `xxxxxxxxxnotexisting` looks off in production code; maybe add a `private static final String CACHE_BUSTER_FIELD = "__pds_registry_cache_buster__"` and use `CACHE_BUSTER_FIELD`
- `rand.nextInt()` can collide; consider using `UUID.randomUUID().toString()` or `System.nanoTime()` instead
- This changes query semantics in a weird way just to influence caching

But I think OpenSearch has a `request_cache=False` option anyway or request an explicit refresh after load if possible.

File: src/main/java/gov/nasa/pds/registry/common/connection/es/SearchImpl.java [hunk 6]
```diff
@@ -28,6 +28,14 @@ public Search buildListLdds(String namespace) {
     this.json = JsonHelper.buildListLddsRequest(namespace);
     return this;
   }
+
+  @Override
+  public Search buildListLddsNoCache(String namespace) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/connection/es/SearchImpl.java:33]: Should we change the name of this function? In `aws/SearchImpl.java` the function `buildListLddsNoCache` adds a cache-busting query, but here in `es/SearchImpl.java`, it's just a normal request.

If that's intentional, fine. But this means the interface name promises behavior that one implementation does not provide. The comment says the cache is disabled, but it's easy to misunderstand later. Maybe rename to something less "absolute", like `buildListLddsFresh(…)`?

File: src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java [hunk 7]
```diff
@@ -17,6 +17,7 @@
 
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.opensearch.client.opensearch.core.bulk.BulkResponseItem;
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/dao/DataLoader.java]: Import unused?

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 8]
```diff
@@ -120,6 +129,28 @@ public void loadOnly(File lddFile, String lddFileName, String namespace, Instant
...
       loader.loadFile(tempEsDataFile);
+
+      // check that the new Ldd is indexed by querying for LDD info 
+      // until the new LDD file name appears in the list of loaded LDDs for this namespace. This is needed to ensure that the new LDD is indexed before any other LDDs are loaded (e.g., from other threads) and potentially overwrite the new LDD with an older one.
+      // TODO: remove this check after we migrate to managed opensearch
+      
+ 
+      LddVersions info = dao.getLddInfoNoCache(namespace);
+
+      /*
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: Please explain commented-out code or erase it (with extreme prejudice)

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 9]
```diff
@@ -2,9 +2,14 @@
 
 import java.io.File;
 import java.time.Instant;
+import java.util.ArrayList;
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: Unused import `ArrayList`?

Also applies to `Collection`, `List`, and `Tuple`, below.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java [hunk 10]
```diff
@@ -88,9 +93,13 @@ public void load(File lddFile, String lddFileName, String namespace) throws Exce
...
     }
 
+    // get date of the current LDD for this namespace in the registry. If there is no LDD for this namespace, use epoch date.
+    // TODO add the test of overwrite yes/no, remove it from where it is not >?
+
+
     // Get information about LDDs already loaded into the registry (for this namespace)
     LddVersions info = dao.getLddInfo(namespace);
-    if (info.files.contains(lddFileName)) {
+    if (info.files.contains(lddFileName) ) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: I'd remove this extra space (marked `␣`) in `if (info.files.contains(lddFileName)␣) {`

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java [hunk 11]
```diff
@@ -88,6 +93,21 @@ public Search buildListLdds(String namespace) {
     this.craftsman.source(new SourceConfig.Builder().filter(new SourceFilter.Builder().includes("date", "attr_name").build()).build());
     return this;
   }
+
+@Override
+public Search buildListLddsNoCache(String namespace) {
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: Is the indentation of this function correct? It looks a bit "left".

File: src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java [hunk 12]
```diff
@@ -52,6 +52,24 @@ public LddVersions getLddInfo(String namespace)
     return client.performRequest(req).lddInfo();
   }
 
+    /**
+   * Get LDD date from data dictionary index in Elasticsearch. Force skip the OpenSearch cache.
```
Comment (nutjob4life) [inline src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java:56]: Javadoc indentation here looks off

**Review by nutjob4life — CHANGES_REQUESTED** [review]
I think this is addressing the right problem, but I’d like to see a few changes before merge; see the comments interspersed.

**Comment by jordanpadams** [comment]
@tloubrieu-jpl can you make some of the updates recommended by @nutjob4life ?

File: src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java (hunk 12 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java:56]: @copilot can you fix this?

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 10 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: @copilot can you fix this?

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 9 above)
Comment (jordanpadams) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: @copilot can you fix this?

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 1 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: ok, I removed the commented lines remaining from tests I've done and could need again, but AI will do it again for me if needed.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 2 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: Let's have a call to clarify that

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 3 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java:136]: ok, good idea, done.

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java (hunk 4 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: That was intentional. I had it at the class level since the method is called multiple times in a row, while we are waiting for the LDD data to be ready in opensearch, and I did not want to re-instanciate the object each time the method is called.

Do you think it is not worth it ?

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java (hunk 5 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: Agreed that looks hacky I'm trying the request_cache parameter now.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 2 above)
Comment (nutjob4life) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: @tloubrieu-jpl Based on the huddle we can resolve this conversation immediately

File: src/main/java/gov/nasa/pds/registry/common/connection/es/SearchImpl.java (hunk 6 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/connection/es/SearchImpl.java:33]: Yes I agree this is confusing but I don't want to spend the extra-time to try to fix an implementation which is not used at all, except for integration test which already work without the need to the nocache specific behavior.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 10 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: thanks

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java (hunk 11 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: Right! It was "left" indeed. I fixed that.

File: src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java (hunk 12 above)
Comment (tloubrieu-jpl) [reply src/main/java/gov/nasa/pds/registry/common/es/dao/dd/DataDictionaryDao.java:56]: Thanks

File: src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java (hunk 4 above)
Comment (nutjob4life) [reply src/main/java/gov/nasa/pds/registry/common/connection/aws/SearchImpl.java]: That'll be fine then @tloubrieu-jpl, leave it as is.

File: src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java (hunk 1 above)
Comment (nutjob4life) [reply src/main/java/gov/nasa/pds/registry/common/es/service/JsonLddLoader.java]: @tloubrieu-jpl you can always add that precise explanation to the commented-out code:
```javascript
/* The following code is commented-out that may be needed again for reason _________________:

      ids.add("insight:Observation_Information/insight:stop_local_mean_solar_time"); 
      List<Tuple> returnedLDDFields = dao.getDataTypes(ids);
      log.info("Returned LDD field: " + returnedLDDFields.toString());
*/
```
Up to you, of course! 😄

**Review by nutjob4life — APPROVED** [review]
Delta: ✓
Discussion: ✓
Disposition: ✓
Approval: ✅
Gratitude: Thanks @tloubrieu-jpl for going through all this!

**Comment by tloubrieu-jpl** [comment]
Thanks Sean for the value added review, I also added a wait/verification step after the LDD is loaded as it was necessary to pass the integration tests.
